src/firstboot/firstboot.c

```diff
@@ -503,7 +505,8 @@ static int process_hostname(void) {
                 return 0;
 
         r = write_string_file(etc_hostname, arg_hostname,
-                              WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_SYNC | WRITE_STRING_FILE_MKDIR_0755);
+                              WRITE_STRING_FILE_CREATE | WRITE_STRING_FILE_SYNC | WRITE_STRING_FILE_MKDIR_0755 |
+                              (arg_overwrite ? WRITE_STRING_FILE_OVERWRITE : 0));
```

Comment:
This sounds like a job for WRITE_STRING_FILE_ATOMIC so that either the old or the new file are in place... And we don't need WRITE_STRING_FILE_OVERWRITE at all then